substrate/client/network/sync/src/strategy/chain_sync.rs

```diff
@@ -1649,13 +1737,34 @@ where
 				return None
 			}
 
-			for (id, peer) in self.peers.iter_mut() {
-				if peer.state.is_available() && peer.common_number >= sync.target_number() {
-					peer.state = PeerSyncState::DownloadingState;
-					let request = sync.next_request();
-					trace!(target: LOG_TARGET, "New StateRequest for {}: {:?}", id, request);
-					self.allowed_requests.clear();
-					return Some((*id, OpaqueStateRequest(Box::new(request))))
+			for peer_id in self.peer_pool.available_peers() {
+				if let Some(peer) = self.peers.get_mut(&peer_id) {
+					if peer.state.is_available() && peer.common_number >= sync.target_number() {
+						if self.peer_pool.try_reserve_peer(&peer_id) {
```

Comment:
Same here, this should not be possible/it shouldn't be necessary to check that a peer is available in two different places.